File: packages/litertlm-native/ios/EngineBridge.swift
import Foundation
import os

/// Phase 1 — LiteRT-LM Engine wrapper + InferenceStateBridge skeleton (iOS parity with Android).
final class EngineBridge {
  private var lifecycle = "unloaded"
  private var engine: Engine?
  private var conversations: [String: Conversation] = [:]
  private var generationTasks: [String: Task<Void, Never>] = [:]
  private let toolApprovalGate = ToolApprovalGate()
  private var lastModelPath: String?
  private var lastBackend = "cpu"
  private var lastCacheDir: String?
  private var persistKvOnHibernate = true
  var onHibernationPolicyChanged: ((Bool) -> Void)?
  private let stateLock = OSAllocatedUnfairLock()

  var onLifecycleChanged: ((String, String) -> Void)?
  var onStreamDelta: ((String, String, String) -> Void)?
  var onMessageComplete: ((String, String) -> Void)?
  var onError: ((String, String) -> Void)?
  var onToolApprovalRequired: ((
    _ conversationId: String,
    _ toolCallId: String,
    _ name: String,
    _ argumentsJson: String,
    _ riskLevel: String
  ) -> Void)?

  func getLifecycle() -> String {
    lifecycle
  }

  func initialize(modelPath: String, backend: String, cacheDir: String?) async throws {
    let normalizedPath = try Self.normalizeRequiredPath(modelPath)
    let normalizedCache = Self.normalizeOptionalPath(cacheDir) ?? Self.defaultCacheDir()

    lastModelPath = normalizedPath
    lastBackend = backend
    lastCacheDir = normalizedCache

    transition("loading")
    do {
      await closeEngineLocked()
      try Self.validateModelFile(at: normalizedPath)

      let parsedBackend = Self.parseBackend(backend)
      let newEngine = try await Self.createEngineOnBackground(
        modelPath: normalizedPath,
        backend: parsedBackend,
        cacheDir: normalizedCache
      )

      withStateLock {
        engine = newEngine
      }
      transition("active")
    } catch {
      await closeEngineLocked()
      transition("error")
      let message = Self.describeInitFailure(error, modelPath: normalizedPath, backend: backend)
      onError?("ENGINE_INIT_FAILED", message)
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: message]
      )
    }
  }

  func warmUp(modelPath: String, backend: String, cacheDir: String?) async throws {
    if lifecycle == "active" || lifecycle == "idle" || lifecycle == "loading" {
      return
    }
    try await initialize(modelPath: modelPath, backend: backend, cacheDir: cacheDir)
  }

  func shutdown() async {
    await closeEngineLocked()
    transition("unloaded")
  }

  func createConversation(
    conversationId: String,
    systemInstruction: String?,
    configJson: String? = nil
  ) async throws {
    let activeEngine = withStateLock { engine }

    guard let activeEngine else {
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "ENGINE_NOT_READY"]
      )
    }
    if lifecycle != "active" && lifecycle != "idle" {
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "ENGINE_NOT_READY"]
      )
    }

    let parsed = ConversationConfigJson.parse(configJson)
    var config = ConversationConfig()
    if let instruction = systemInstruction?.trimmingCharacters(in: .whitespacesAndNewlines),
       !instruction.isEmpty
    {
      config = ConversationConfig(systemMessage: Message(instruction, role: .system))
    }

    if parsed.enableBuiltinTools {
      BuiltinToolContext.configure(
        conversationId: conversationId,
        approvalGate: toolApprovalGate,
        onApprovalRequired: { [weak self] convId, toolCallId, name, argumentsJson, riskLevel in
          self?.onToolApprovalRequired?(convId, toolCallId, name, argumentsJson, riskLevel)
        }
      )
      config = ConversationConfig(
        systemMessage: config.systemMessage,
        initialMessages: config.initialMessages,
        tools: BuiltinToolContext.builtinTools(),
        samplerConfig: parsed.sampler
      )
    } else if let sampler = parsed.sampler {
      config = ConversationConfig(
        systemMessage: config.systemMessage,
        initialMessages: config.initialMessages,
        tools: config.tools,
        samplerConfig: sampler
      )
    }

    let conversation = try await activeEngine.createConversation(with: config)
    withStateLock {
      conversations[conversationId] = conversation
    }
  }

  func approveToolCall(conversationId: String, toolCallId: String, approved: Bool) {
    toolApprovalGate.resolve(toolCallId: toolCallId, approved: approved)
  }

  func rejectToolCall(conversationId: String, toolCallId: String, reason: String?) {
    toolApprovalGate.resolve(toolCallId: toolCallId, approved: false)
  }

  func abortGeneration(conversationId: String) {
    generationTasks.removeValue(forKey: conversationId)?.cancel()
    if let conversation = withStateLock({ conversations[conversationId] }) {
      try? conversation.cancel()
    }
    onError?("GENERATION_ABORTED", "Generation aborted")
  }

  func closeConversation(conversationId: String) async {
    withStateLock {
      _ = conversations.removeValue(forKey: conversationId)
    }
  }

  func sendMessage(
    conversationId: String,
    text: String,
    extraContext: [String: Any] = [:]
  ) {
    if lifecycle != "active" && lifecycle != "idle" {
      onError?("ENGINE_NOT_READY", "Engine lifecycle=\(lifecycle)")
      return
    }

    let conversation = withStateLock { conversations[conversationId] }

    guard let conversation else {
      onError?("CONVERSATION_NOT_FOUND", conversationId)
      return
    }

    let batcher = TokenBatcher { [weak self] delta, kind in
      self?.onStreamDelta?(conversationId, delta, kind)
    }

    let task = Task {
      var fullResponse = ""
      do {
        for try await chunk in conversation.sendMessageStream(Message(text), extraContext: extraContext) {
          if Task.isCancelled {
            throw CancellationError()
          }
          for (_, value) in chunk.channels where !value.isEmpty {
            batcher.append(value, kind: "thinking")
          }
          let tokenText = chunk.toString
          if !tokenText.isEmpty {
            batcher.append(tokenText, kind: "token")
            fullResponse += tokenText
          }
        }
        batcher.flush()
        onMessageComplete?(conversationId, fullResponse)
      } catch is CancellationError {
        batcher.flush()
        onError?("GENERATION_ABORTED", "Generation aborted")
      } catch {
        onError?("GENERATION_FAILED", error.localizedDescription)
      }
      generationTasks.removeValue(forKey: conversationId)
    }
    generationTasks[conversationId] = task
  }

  func enterIdle() {
    if lifecycle == "active" {
      transition("idle")
    }
  }

  func setHibernationPolicy(persistKvOnHibernate: Bool, hibernateOnMemoryWarning: Bool) {
    self.persistKvOnHibernate = persistKvOnHibernate
    onHibernationPolicyChanged?(hibernateOnMemoryWarning)
  }

  func hibernate(conversationIds: [String]? = nil) async {
    transition("hibernating")
    if persistKvOnHibernate {
      let ids: [String]
      if let conversationIds, !conversationIds.isEmpty {
        ids = conversationIds
      } else {
        ids = withStateLock { Array(conversations.keys) }
      }
      for conversationId in ids {
        _ = persistSession(conversationId: conversationId, messageCount: 0)
      }
    }
    await closeEngineLocked()
    transition("hibernated")
  }

  func persistSession(conversationId: String, messageCount: Int = 0) -> [String: Any] {
    let cacheDir = lastCacheDir ?? Self.defaultCacheDir()
    let meta =
      (try? SessionSnapshotStore.writeMeta(
        cacheDir: cacheDir,
        conversationId: conversationId,
        messageCount: messageCount,
        usedNativeKvSerialize: false
      ))
    let snapshotPath = SessionSnapshotStore.snapshotFile(cacheDir: cacheDir, conversationId: conversationId).path
    return [
      "conversationId": conversationId,
      "snapshotPath": snapshotPath,
      "snapshotBytes": 0,
      "usedNativeKvSerialize": meta?.usedNativeKvSerialize ?? false,
    ]
  }

  func restoreSession(conversationId: String) async throws -> [String: Any] {
    let cacheDir = lastCacheDir ?? Self.defaultCacheDir()
    let meta = SessionSnapshotStore.readMeta(cacheDir: cacheDir, conversationId: conversationId)
    let restoredFrom = SessionSnapshotStore.restoredFrom(meta)

    transition("restoring")
    defer {
      if lifecycle == "restoring" {
        transition(engine != nil ? "active" : "idle")
      }
    }

    let hasConversation = withStateLock { conversations[conversationId] != nil }
    let hasEngine = withStateLock { engine != nil }

    if !hasConversation && hasEngine {
      try await createConversation(conversationId: conversationId, systemInstruction: nil, configJson: nil)
    }
    return [
      "conversationId": conversationId,
      "restoredFrom": restoredFrom,
      "prefillSkippedTokens": restoredFrom == "kv_snapshot" ? (meta?.messageCount ?? 0) : 0,
    ]
  }

  func deleteSessionSnapshot(conversationId: String) {
    let cacheDir = lastCacheDir ?? Self.defaultCacheDir()
    SessionSnapshotStore.deleteSnapshot(cacheDir: cacheDir, conversationId: conversationId)
  }

  private func transition(_ next: String) {
    let from = lifecycle
    lifecycle = next
    DispatchQueue.main.async { [weak self] in
      self?.onLifecycleChanged?(from, next)
    }
  }

  private func closeEngineLocked() async {
    let hadEngine = withStateLock { () -> Bool in
      let hadEngine = engine != nil
      conversations.removeAll()
      engine = nil
      return hadEngine
    }

    // Let the Engine actor deinit and release native handles before re-create.
    if hadEngine {
      try? await Task.sleep(nanoseconds: 150_000_000)
    }
  }

  private func withStateLock<T>(_ body: () -> T) -> T {
    stateLock.withLock(body)
  }

  private static func createEngineOnBackground(
    modelPath: String,
    backend: Backend,
    cacheDir: String
  ) async throws -> Engine {
    try await Task.detached(priority: .userInitiated) {
      let engineConfig = try EngineConfig(
        modelPath: modelPath,
        backend: backend,
        cacheDir: cacheDir
      )
      let newEngine = Engine(engineConfig: engineConfig)
      try await newEngine.initialize()
      return newEngine
    }.value
  }

  private static func validateModelFile(at path: String) throws {
    var isDirectory: ObjCBool = false
    guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory) else {
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "MODEL_NOT_FOUND: \(path)"]
      )
    }
    if isDirectory.boolValue {
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "MODEL_NOT_FOUND: path is a directory: \(path)"]
      )
    }

    let size = (try? FileManager.default.attributesOfItem(atPath: path)[.size] as? NSNumber)?
      .int64Value ?? 0
    if size <= 0 {
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "MODEL_NOT_FOUND: model file is empty"]
      )
    }
  }

  private static func describeInitFailure(_ error: Error, modelPath: String, backend: String) -> String {
    let base = (error as NSError).userInfo[NSLocalizedDescriptionKey] as? String
      ?? error.localizedDescription
    if base.contains("MODEL_NOT_FOUND") {
      return base
    }
    if base.contains("Failed to create engine") {
      return """
        \(base) (backend=\(backend)). \
        Check model path and free memory. Path: \(modelPath)
        """
    }
    return base
  }

  private static func defaultCacheDir() -> String {
    FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.path
      ?? NSTemporaryDirectory()
  }

  private static func parseBackend(_ backend: String) -> Backend {
    switch backend.lowercased() {
    case "gpu":
      return .gpu
    default:
      return .cpu()
    }
  }

  private static func normalizeRequiredPath(_ path: String) throws -> String {
    let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
      throw NSError(
        domain: "LitertlmNative",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "MODEL_NOT_FOUND: modelPath is empty"]
      )
    }
    return normalizeOptionalPath(trimmed) ?? trimmed
  }

  private static func normalizeOptionalPath(_ path: String?) -> String? {
    guard let path, !path.isEmpty else {
      return nil
    }
    if path.hasPrefix("file://") {
      return URL(string: path)?.path ?? String(path.dropFirst("file://".count))
    }
    return path
  }

  static func parseExtraContext(_ extraContextJson: String?) -> [String: Any] {
    guard let extraContextJson, !extraContextJson.isEmpty,
          let data = extraContextJson.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else {
      return [:]
    }
    return json
  }
}

File: packages/litertlm-native/android/src/main/java/expo/modules/litertlmnative/EngineBridge.kt
package expo.modules.litertlmnative

import android.content.Context
import com.google.ai.edge.litertlm.Backend
import com.google.ai.edge.litertlm.Contents
import com.google.ai.edge.litertlm.Conversation
import com.google.ai.edge.litertlm.ConversationConfig
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import com.google.ai.edge.litertlm.Message
import com.google.ai.edge.litertlm.tool
import com.google.gson.JsonParser
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap

/**
 * Phase 1 — LiteRT-LM Engine wrapper + InferenceStateBridge skeleton.
 */
class EngineBridge(private val context: Context) {
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
  private val bridgeMutex = Mutex()
  private val conversations = ConcurrentHashMap<String, Conversation>()
  private val generationJobs = ConcurrentHashMap<String, Job>()
  private val toolApprovalGate = ToolApprovalGate()
  private var lifecycle: String = "unloaded"
  private var engine: Engine? = null
  private var lastModelPath: String? = null
  private var lastBackend: String = "cpu"
  private var lastCacheDir: String? = null
  @Volatile
  private var persistKvOnHibernate: Boolean = true

  var onLifecycleChanged: ((from: String, to: String) -> Unit)? = null
  var onStreamDelta: ((conversationId: String, delta: String, kind: String) -> Unit)? = null
  var onMessageComplete: ((conversationId: String, content: String) -> Unit)? = null
  var onError: ((code: String, message: String) -> Unit)? = null
  var onToolApprovalRequired: ((
    conversationId: String,
    toolCallId: String,
    name: String,
    argumentsJson: String,
    riskLevel: String,
  ) -> Unit)? = null

  fun getLifecycle(): String = lifecycle

  suspend fun initialize(modelPath: String, backend: String, cacheDir: String?) {
    bridgeMutex.withLock {
      initializeLocked(modelPath, backend, cacheDir)
    }
  }

  private suspend fun initializeLocked(modelPath: String, backend: String, cacheDir: String?) {
    lastModelPath = normalizeFilesystemPath(modelPath) ?: modelPath
    lastBackend = backend
    lastCacheDir = normalizeFilesystemPath(cacheDir) ?: context.cacheDir.path

    transition("loading")
    try {
      withContext(Dispatchers.IO) {
        closeEngineLocked()
        val engineConfig =
          EngineConfig(
            modelPath = lastModelPath!!,
            backend = parseBackend(backend),
            cacheDir = lastCacheDir,
          )
        val newEngine = Engine(engineConfig)
        newEngine.initialize()
        engine = newEngine
      }
      transition("active")
    } catch (error: Exception) {
      closeEngineLocked()
      transition("error")
      onError?.invoke("ENGINE_INIT_FAILED", error.message ?: "initialize failed")
      throw error
    }
  }

  suspend fun warmUp(modelPath: String, backend: String, cacheDir: String?) {
    bridgeMutex.withLock {
      if (lifecycle == "active" || lifecycle == "idle" || lifecycle == "loading") {
        return
      }
      initializeLocked(modelPath, backend, cacheDir)
    }
  }

  suspend fun shutdown() {
    bridgeMutex.withLock {
      withContext(Dispatchers.IO) {
        closeEngineLocked()
      }
      transition("unloaded")
    }
  }

  suspend fun createConversation(
    conversationId: String,
    systemInstruction: String?,
    configJson: String? = null,
  ) {
    bridgeMutex.withLock {
      createConversationLocked(conversationId, systemInstruction, configJson)
    }
  }

  fun approveToolCall(_conversationId: String, toolCallId: String, approved: Boolean) {
    toolApprovalGate.resolve(toolCallId, approved)
  }

  fun rejectToolCall(_conversationId: String, toolCallId: String, _reason: String?) {
    toolApprovalGate.resolve(toolCallId, false)
  }

  fun abortGeneration(conversationId: String) {
    generationJobs.remove(conversationId)?.cancel()
    runCatching { conversations[conversationId]?.cancelProcess() }
    onError?.invoke("GENERATION_ABORTED", "Generation aborted")
  }

  suspend fun closeConversation(conversationId: String) {
    bridgeMutex.withLock {
      withContext(Dispatchers.IO) {
        conversations.remove(conversationId)?.close()
      }
    }
  }

  fun sendMessage(
    conversationId: String,
    text: String,
    extraContext: Map<String, Any> = emptyMap(),
  ) {
    if (lifecycle != "active" && lifecycle != "idle") {
      onError?.invoke("ENGINE_NOT_READY", "Engine lifecycle=$lifecycle")
      return
    }

    val conversation =
      conversations[conversationId]
        ?: run {
          onError?.invoke("CONVERSATION_NOT_FOUND", conversationId)
          return
        }

    val batcher =
      TokenBatcher(
        onFlush = { delta, kind ->
          onStreamDelta?.invoke(conversationId, delta, kind)
        },
      )

    val job =
      scope.launch {
        var fullResponse = ""
        try {
          conversation
            .sendMessageAsync(text, extraContext = extraContext)
            .catch { error ->
              if (error is CancellationException) {
                throw error
              }
              onError?.invoke("GENERATION_FAILED", error.message ?: "sendMessage failed")
            }
            .collect { message ->
              for ((delta, kind) in extractStreamParts(message)) {
                batcher.append(delta, kind)
                if (kind == "token") {
                  fullResponse += delta
                }
              }
            }
          batcher.flush()
          onMessageComplete?.invoke(conversationId, fullResponse)
        } catch (error: CancellationException) {
          batcher.flush()
          onError?.invoke("GENERATION_ABORTED", "Generation aborted")
        } catch (error: Exception) {
          onError?.invoke("GENERATION_FAILED", error.message ?: "sendMessage failed")
        } finally {
          generationJobs.remove(conversationId)
        }
      }
    generationJobs[conversationId] = job
  }

  suspend fun enterIdle() {
    bridgeMutex.withLock {
      if (lifecycle == "active") {
        transition("idle")
      }
    }
  }

  fun setHibernationPolicy(persistKvOnHibernate: Boolean, hibernateOnMemoryWarning: Boolean) {
    this.persistKvOnHibernate = persistKvOnHibernate
    onHibernationPolicyChanged?.invoke(hibernateOnMemoryWarning)
  }

  var onHibernationPolicyChanged: ((hibernateOnMemoryWarning: Boolean) -> Unit)? = null

  suspend fun hibernate(conversationIds: List<String>? = null) {
    bridgeMutex.withLock {
      transition("hibernating")
      withContext(Dispatchers.IO) {
        if (persistKvOnHibernate) {
          val ids =
            conversationIds?.takeIf { it.isNotEmpty() }
              ?: conversations.keys().toList()
          for (conversationId in ids) {
            runCatching { persistSessionLocked(conversationId, messageCount = 0) }
          }
        }
        closeEngineLocked()
      }
      transition("hibernated")
    }
  }

  suspend fun persistSession(conversationId: String, messageCount: Int = 0): Map<String, Any> {
    return bridgeMutex.withLock {
      persistSessionLocked(conversationId, messageCount)
    }
  }

  private fun persistSessionLocked(conversationId: String, messageCount: Int): Map<String, Any> {
    val cacheDir = lastCacheDir ?: context.cacheDir.path
    val meta =
      SessionSnapshotStore.writeMeta(
        cacheDir = cacheDir,
        conversationId = conversationId,
        messageCount = messageCount,
        usedNativeKvSerialize = false,
      )
    val snapshotPath = SessionSnapshotStore.snapshotFile(cacheDir, conversationId).path
    return mapOf(
      "conversationId" to conversationId,
      "snapshotPath" to snapshotPath,
      "snapshotBytes" to 0,
      "usedNativeKvSerialize" to meta.usedNativeKvSerialize,
    )
  }

  suspend fun restoreSession(conversationId: String): Map<String, Any> {
    val cacheDir = lastCacheDir ?: context.cacheDir.path
    val meta = SessionSnapshotStore.readMeta(cacheDir, conversationId)
    val restoredFrom = SessionSnapshotStore.restoredFrom(meta)

    transition("restoring")
    try {
      bridgeMutex.withLock {
        if (!conversations.containsKey(conversationId) && engine != null) {
          createConversationLocked(conversationId, null, null)
        }
      }
      return mapOf(
        "conversationId" to conversationId,
        "restoredFrom" to restoredFrom,
        "prefillSkippedTokens" to if (restoredFrom == "kv_snapshot") (meta?.messageCount ?: 0) else 0,
      )
    } finally {
      if (lifecycle == "restoring") {
        transition(if (engine != null) "active" else "idle")
      }
    }
  }

  suspend fun deleteSessionSnapshot(conversationId: String) {
    val cacheDir = lastCacheDir ?: context.cacheDir.path
    withContext(Dispatchers.IO) {
      SessionSnapshotStore.deleteSnapshot(cacheDir, conversationId)
    }
  }

  private suspend fun createConversationLocked(
    conversationId: String,
    systemInstruction: String?,
    configJson: String?,
  ) {
    val activeEngine = engine ?: throw IllegalStateException("ENGINE_NOT_READY")
    if (lifecycle != "active" && lifecycle != "idle") {
      throw IllegalStateException("ENGINE_NOT_READY")
    }

    val parsed = ConversationConfigJson.parse(configJson)

    withContext(Dispatchers.IO) {
      conversations.remove(conversationId)?.close()

      val tools =
        if (parsed.enableBuiltinTools) {
          listOf(
            tool(
              DeviceToolSet(
                context = context,
                conversationId = conversationId,
                approvalGate = toolApprovalGate,
                onApprovalRequired = { convId, toolCallId, name, argumentsJson, riskLevel ->
                  onToolApprovalRequired?.invoke(
                    convId,
                    toolCallId,
                    name,
                    argumentsJson,
                    riskLevel,
                  )
                },
              ),
            ),
          )
        } else {
          emptyList()
        }

      val config =
        ConversationConfig(
          systemInstruction = systemInstruction?.takeIf { it.isNotBlank() }?.let { Contents.of(it) },
          tools = tools,
          samplerConfig = parsed.sampler,
          automaticToolCalling = parsed.automaticToolCalling,
        )
      conversations[conversationId] = activeEngine.createConversation(config)
    }
  }

  private suspend fun transition(next: String) {
    val from = lifecycle
    lifecycle = next
    withContext(Dispatchers.Main) {
      onLifecycleChanged?.invoke(from, next)
    }
  }

  private fun closeEngineLocked() {
    for ((_, conversation) in conversations) {
      runCatching { conversation.close() }
    }
    conversations.clear()
    engine?.close()
    engine = null
  }

  private fun parseBackend(backend: String): Backend =
    when (backend.lowercase()) {
      "gpu" -> Backend.GPU()
      "npu" ->
        Backend.NPU(nativeLibraryDir = context.applicationInfo.nativeLibraryDir)
      else -> Backend.CPU()
    }

  private fun normalizeFilesystemPath(path: String?): String? {
    if (path.isNullOrBlank()) {
      return null
    }
    return when {
      path.startsWith("file://") -> path.removePrefix("file://")
      else -> path
    }
  }

  private fun extractStreamParts(message: Message): List<Pair<String, String>> {
    val parts = mutableListOf<Pair<String, String>>()
    for ((_, value) in message.channels) {
      if (value.isNotEmpty()) {
        parts.add(value to "thinking")
      }
    }
    val text = message.toString()
    if (text.isNotEmpty()) {
      parts.add(text to "token")
    }
    return parts
  }

  companion object {
    fun parseExtraContext(extraContextJson: String?): Map<String, Any> {
      if (extraContextJson.isNullOrBlank()) {
        return emptyMap()
      }
      val json = JsonParser.parseString(extraContextJson).asJsonObject
      val result = mutableMapOf<String, Any>()
      for ((key, value) in json.entrySet()) {
        when {
          value.isJsonPrimitive && value.asJsonPrimitive.isBoolean ->
            result[key] = value.asBoolean
          value.isJsonPrimitive && value.asJsonPrimitive.isNumber ->
            result[key] = value.asNumber
          value.isJsonPrimitive ->
            result[key] = value.asString
        }
      }
      return result
    }
  }
}
